eth/downloader/downloader_test.go

```diff
@@ -1541,3 +1530,42 @@ func testFakedSyncProgress(t *testing.T, protocol int, mode SyncMode) {
 		t.Fatalf("Final progress mismatch: have %v/%v/%v, want 0-%v/%v/%v", origin, current, latest, targetBlocks, targetBlocks, targetBlocks)
 	}
 }
+
+// This test reproduces an issue where unexpected deliveries would
+// block indefinitely if they arrived at the right time.
+func TestDeliverHeadersHang62(t *testing.T) {
```

Comment:
PS: I know the code being tested is currently generic between 62+, but adding the tests for all of them doesn't hurt for now and is more future proof with respect to protocol modifications (e.g. don't know if LES will require or not slight modifications later for light sync, so it may diverge at that point, possibly missing out on test coverage).